browser/ui/commands/accelerator_service.cc

```diff
@@ -59,32 +61,78 @@ AcceleratorService::AcceleratorService(PrefService* pref_service,
 AcceleratorService::~AcceleratorService() = default;
 
 void AcceleratorService::Initialize() {
-  auto accelerators = pref_manager_.GetAccelerators();
-  if (accelerators.size() == 0) {
-    for (const auto& [command_id, default_accelerators] :
-         default_accelerators_) {
-      for (const auto& default_accelerator : default_accelerators) {
-        pref_manager_.AddAccelerator(command_id,
-                                     ToCodesString(default_accelerator));
-      }
-    }
-    accelerators = pref_manager_.GetAccelerators();
-  }
+  accelerators_ = pref_manager_.GetAccelerators();
+  UpdateDefaultAccelerators();
+
+  accelerators_ = pref_manager_.GetAccelerators();
 
+  // Include commands in the table which don't have any accelerators.
   auto commands = GetCommands();
   for (const auto& command : commands) {
-    accelerators_.insert({command, {}});
+    if (!accelerators_.contains(command)) {
+      accelerators_[command] = {};
+    }
+  }
+}
 
-    auto it = accelerators.find(command);
-    if (it == accelerators.end()) {
+void AcceleratorService::UpdateDefaultAccelerators() {
+  auto old_defaults = pref_manager_.GetDefaultAccelerators();
+
+  Accelerators added;
+  Accelerators removed;
+
+  // Handle new accelerators, and removed accelerators.
+  for (const auto& command : default_accelerators_) {
+    const auto& new_accelerators = command.second;
+    const auto& old_accelerators = old_defaults[command.first];
+
+    // Note all the added accelerators.
+    std::copy_if(new_accelerators.begin(), new_accelerators.end(),
+                 std::back_inserter(added[command.first]),
+                 [&old_accelerators](const auto& accelerator) {
+                   return !base::Contains(old_accelerators, accelerator);
+                 });
+
+    // Note all the removed accelerators.
+    std::copy_if(old_accelerators.begin(), old_accelerators.end(),
+                 std::back_inserter(removed[command.first]),
+                 [&new_accelerators](const auto& accelerator) {
+                   return !base::Contains(new_accelerators, accelerator);
+                 });
```

Comment:
Yeah, I think the current defaults all have 0-5 shortcuts (and only one has 5) so it might be a bit worse on performance. That said, it definitely reads nicely :laughing: